File: autogpt_platform/copilot-bot/src/platform-api.ts
/**
 * Client for the AutoGPT Platform Linking & Chat APIs.
 *
 * Two independent linking flows:
 *   - SERVER: /resolve, /tokens, /tokens/{t}/status — claimed server context
 *   - USER:   /resolve-user, /user-tokens — 1:1 DM context
 *
 * Chat endpoints (/chat/session, /chat/stream) accept either context — pass
 * a serverId for server messages, omit for DMs.
 */

const DEFAULT_TIMEOUT_MS = 30_000;
// Idle timeout: abort only if no data arrives for this long. Backend sends
// `: keepalive\n\n` every 30s, so 90s gives 3 missed keepalives of headroom.
// CoPilot turns can legitimately take many minutes — a hard deadline would kill
// long-running tool calls mid-flight.
const SSE_IDLE_TIMEOUT_MS = 90_000;

export class PlatformAPIError extends Error {
  constructor(
    public readonly status: number,
    message: string,
  ) {
    super(message);
    this.name = "PlatformAPIError";
  }
}

export interface ResolveResult {
  linked: boolean;
}

export interface LinkTokenResult {
  token: string;
  expires_at: string;
  link_url: string;
}

export interface LinkTokenStatus {
  status: "pending" | "linked" | "expired";
}

export class PlatformAPI {
  private readonly botApiKey: string;

  constructor(private readonly baseUrl: string) {
    const key = process.env.PLATFORM_BOT_API_KEY;
    if (!key) {
      throw new Error(
        "PLATFORM_BOT_API_KEY is required. Set it in your .env file.",
      );
    }
    this.botApiKey = key;
  }

  /** Check if a server has a PlatformLink (anyone in it can chat). */
  async resolveServer(
    platform: string,
    platformServerId: string,
  ): Promise<ResolveResult> {
    return this.postJson("/api/platform-linking/resolve", {
      platform: platform.toUpperCase(),
      platform_server_id: platformServerId,
    });
  }

  /** Check if an individual has a PlatformUserLink (their DMs are linked). */
  async resolveUser(
    platform: string,
    platformUserId: string,
  ): Promise<ResolveResult> {
    return this.postJson("/api/platform-linking/resolve-user", {
      platform: platform.toUpperCase(),
      platform_user_id: platformUserId,
    });
  }

  /**
   * Create a SERVER link token. platformUserId is the user claiming ownership.
   */
  async createLinkToken(params: {
    platform: string;
    platformServerId: string;
    platformUserId: string;
    platformUsername?: string;
    serverName?: string;
    channelId?: string;
  }): Promise<LinkTokenResult> {
    return this.postJson("/api/platform-linking/tokens", {
      platform: params.platform.toUpperCase(),
      platform_server_id: params.platformServerId,
      platform_user_id: params.platformUserId,
      platform_username: params.platformUsername,
      server_name: params.serverName,
      channel_id: params.channelId,
    });
  }

  /** Create a USER (DM) link token for an individual. */
  async createUserLinkToken(params: {
    platform: string;
    platformUserId: string;
    platformUsername?: string;
  }): Promise<LinkTokenResult> {
    return this.postJson("/api/platform-linking/user-tokens", {
      platform: params.platform.toUpperCase(),
      platform_user_id: params.platformUserId,
      platform_username: params.platformUsername,
    });
  }

  /** Check if a link token has been consumed. Works for both SERVER and USER tokens. */
  async getLinkTokenStatus(token: string): Promise<LinkTokenStatus> {
    const res = await fetch(
      `${this.baseUrl}/api/platform-linking/tokens/${encodeURIComponent(token)}/status`,
      {
        headers: this.headers(),
        signal: AbortSignal.timeout(DEFAULT_TIMEOUT_MS),
      },
    );
    if (!res.ok) throw new PlatformAPIError(res.status, await res.text());
    return res.json();
  }

  /**
   * Create a new CoPilot session. Pass `platformServerId` for server context
   * (session owned by server owner); omit for DM context (owned by the user).
   */
  async createChatSession(params: {
    platform: string;
    platformUserId: string;
    platformServerId?: string;
  }): Promise<string> {
    const data = await this.postJson<{ session_id: string }>(
      "/api/platform-linking/chat/session",
      {
        platform: params.platform.toUpperCase(),
        platform_server_id: params.platformServerId,
        platform_user_id: params.platformUserId,
        message: "session_init",
      },
    );
    return data.session_id;
  }

  /**
   * Stream a chat message. Same context rules as createChatSession —
   * include platformServerId for server messages, omit for DMs.
   */
  async *streamChat(params: {
    platform: string;
    platformUserId: string;
    platformServerId?: string;
    message: string;
    sessionId?: string;
  }): AsyncGenerator<string> {
    const abort = new AbortController();
    let idleTimer: ReturnType<typeof setTimeout> | null = null;
    const resetIdleTimer = () => {
      if (idleTimer) clearTimeout(idleTimer);
      idleTimer = setTimeout(() => abort.abort(), SSE_IDLE_TIMEOUT_MS);
    };
    resetIdleTimer();

    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}/api/platform-linking/chat/stream`, {
        method: "POST",
        headers: { ...this.headers(), Accept: "text/event-stream" },
        body: JSON.stringify({
          platform: params.platform.toUpperCase(),
          platform_server_id: params.platformServerId,
          platform_user_id: params.platformUserId,
          message: params.message,
          session_id: params.sessionId,
        }),
        signal: abort.signal,
      });
    } catch (err) {
      if (idleTimer) clearTimeout(idleTimer);
      throw err;
    }

    if (!res.ok) {
      if (idleTimer) clearTimeout(idleTimer);
      throw new PlatformAPIError(res.status, await res.text());
    }
    if (!res.body) {
      if (idleTimer) clearTimeout(idleTimer);
      throw new PlatformAPIError(0, "No response body for SSE stream");
    }

    const decoder = new TextDecoder();
    const reader = res.body.getReader();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        resetIdleTimer();
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          if (!line.startsWith("data: ")) continue;
          const data = line.slice(6).trim();
          if (data === "[DONE]") return;
          try {
            const parsed = JSON.parse(data) as Record<string, unknown>;
            if (parsed.type === "text-delta" && parsed.delta) {
              yield parsed.delta as string;
            } else if (parsed.type === "error" && parsed.content) {
              yield `Error: ${parsed.content as string}`;
            }
          } catch {
            // Non-JSON line — skip
          }
        }
      }
    } finally {
      if (idleTimer) clearTimeout(idleTimer);
      reader.releaseLock();
    }
  }

  private async postJson<T = unknown>(
    path: string,
    body: Record<string, unknown>,
  ): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(DEFAULT_TIMEOUT_MS),
    });
    if (!res.ok) throw new PlatformAPIError(res.status, await res.text());
    return res.json();
  }

  private headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      "X-Bot-API-Key": this.botApiKey,
    };
  }
}

File: autogpt_platform/backend/backend/copilot/bot/platform_api.py
import json
import logging
from dataclasses import dataclass
from typing import AsyncGenerator, Callable, Optional

import aiohttp

from .config import AUTOGPT_API_URL, PLATFORM_BOT_API_KEY, SSE_IDLE_TIMEOUT

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT = aiohttp.ClientTimeout(total=30)


class PlatformAPIError(Exception):
    def __init__(self, status: int, message: str):
        self.status = status
        super().__init__(f"Platform API error {status}: {message}")


@dataclass
class ResolveResult:
    linked: bool


@dataclass
class LinkTokenResult:
    token: str
    link_url: str
    expires_at: str


class PlatformAPI:
    def __init__(
        self,
        base_url: str = AUTOGPT_API_URL,
        api_key: str = PLATFORM_BOT_API_KEY,
    ):
        self._base_url = base_url.rstrip("/")
        self._headers = {
            "Content-Type": "application/json",
            "X-Bot-API-Key": api_key,
        }
        self._session: Optional[aiohttp.ClientSession] = None

    async def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers=self._headers, timeout=DEFAULT_TIMEOUT
            )
        return self._session

    async def close(self) -> None:
        if self._session and not self._session.closed:
            await self._session.close()

    async def _request(
        self, method: str, path: str, json: Optional[dict] = None
    ) -> dict:
        session = await self._get_session()
        url = f"{self._base_url}{path}"
        async with session.request(method, url, json=json) as resp:
            if resp.status >= 400:
                body = await resp.text()
                raise PlatformAPIError(resp.status, body)
            return await resp.json()

    async def resolve_server(
        self, platform: str, platform_server_id: str
    ) -> ResolveResult:
        data = await self._request(
            "POST",
            "/api/platform-linking/resolve",
            json={
                "platform": platform.upper(),
                "platform_server_id": platform_server_id,
            },
        )
        return ResolveResult(linked=data.get("linked", False))

    async def resolve_user(self, platform: str, platform_user_id: str) -> ResolveResult:
        data = await self._request(
            "POST",
            "/api/platform-linking/resolve-user",
            json={
                "platform": platform.upper(),
                "platform_user_id": platform_user_id,
            },
        )
        return ResolveResult(linked=data.get("linked", False))

    async def create_link_token(
        self,
        platform: str,
        platform_server_id: str,
        platform_user_id: str,
        platform_username: str,
        server_name: str,
        channel_id: str = "",
    ) -> LinkTokenResult:
        data = await self._request(
            "POST",
            "/api/platform-linking/tokens",
            json={
                "platform": platform.upper(),
                "platform_server_id": platform_server_id,
                "platform_user_id": platform_user_id,
                "platform_username": platform_username,
                "server_name": server_name,
                "channel_id": channel_id,
            },
        )
        return LinkTokenResult(
            token=data["token"],
            link_url=data["link_url"],
            expires_at=data["expires_at"],
        )

    async def create_user_link_token(
        self,
        platform: str,
        platform_user_id: str,
        platform_username: str,
    ) -> LinkTokenResult:
        data = await self._request(
            "POST",
            "/api/platform-linking/user-tokens",
            json={
                "platform": platform.upper(),
                "platform_user_id": platform_user_id,
                "platform_username": platform_username,
            },
        )
        return LinkTokenResult(
            token=data["token"],
            link_url=data["link_url"],
            expires_at=data["expires_at"],
        )

    async def stream_chat(
        self,
        platform: str,
        platform_user_id: str,
        message: str,
        session_id: Optional[str] = None,
        platform_server_id: Optional[str] = None,
        on_session_id: Optional[Callable[[str], None]] = None,
    ) -> AsyncGenerator[str, None]:
        """Stream a chat response. If session_id is None, the backend creates one
        and returns it via the X-Session-Id response header — on_session_id() is
        called with that value so the caller can cache it for subsequent messages.
        """
        session = await self._get_session()
        url = f"{self._base_url}/api/platform-linking/chat/stream"
        body: dict = {
            "platform": platform.upper(),
            "platform_user_id": platform_user_id,
            "message": message,
        }
        if session_id:
            body["session_id"] = session_id
        if platform_server_id:
            body["platform_server_id"] = platform_server_id

        timeout = aiohttp.ClientTimeout(total=0, sock_read=SSE_IDLE_TIMEOUT)
        async with session.post(url, json=body, timeout=timeout) as resp:
            if resp.status >= 400:
                error_body = await resp.text()
                raise PlatformAPIError(resp.status, error_body)

            # Backend returns the (possibly new) session ID in this header
            returned_session_id = resp.headers.get("X-Session-Id")
            if returned_session_id and on_session_id:
                on_session_id(returned_session_id)

            async for raw_line in resp.content:
                line = raw_line.decode("utf-8", errors="replace").rstrip("\n\r")

                if not line or line.startswith(":"):
                    continue

                if not line.startswith("data: "):
                    continue

                payload = line[6:]
                if payload == "[DONE]":
                    return

                try:
                    event = json.loads(payload)
                except (json.JSONDecodeError, ValueError):
                    continue

                event_type = event.get("type", "")
                if event_type == "text-delta":
                    delta = event.get("delta", "")
                    if delta:
                        yield delta
                elif event_type == "error":
                    error_msg = event.get("content", "Unknown error")
                    logger.error(f"SSE error from backend: {error_msg}")
                    yield f"\n[Error: {error_msg}]"
                    return
